pkg/cloud/openstack/openstack_test.go

```diff
@@ -31,16 +39,132 @@ func TestResourcesRenderingSmoke(t *testing.T) {
 
 	for _, tc := range tc {
 		t.Run(tc.name, func(t *testing.T) {
+			g := NewWithT(t)
 			assets, err := NewProviderAssets(tc.config)
-			if tc.initErrMsg != "" {
-				assert.EqualError(t, err, tc.initErrMsg)
+			if tc.errMsg != "" {
+				g.Expect(err).Should(MatchError(tc.errMsg))
 				return
 			} else {
-				assert.NoError(t, err)
+				g.Expect(err).ToNot(HaveOccurred())
 			}
 
 			resources := assets.GetRenderedResources()
-			assert.Len(t, resources, 2)
+			g.Expect(resources).Should(HaveLen(1))
+		})
+	}
+}
+
+func makeInfrastructureResource(platform configv1.PlatformType) *configv1.Infrastructure {
+	return &configv1.Infrastructure{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+		Status: configv1.InfrastructureStatus{
+			PlatformStatus: &configv1.PlatformStatus{
+				Type: platform,
+			},
+		},
+		Spec: configv1.InfrastructureSpec{
+			CloudConfig: configv1.ConfigMapFileReference{
+				Name: infraCloudConfName,
+				Key:  infraCloudConfKey,
+			},
+			PlatformSpec: configv1.PlatformSpec{
+				Type: platform,
+			},
+		},
+	}
+}
+
+func TestCloudConfigTransformer(t *testing.T) {
+
+	tc := []struct {
+		name   string
+		source string
+		infra  *configv1.Infrastructure
+		errMsg string
+	}{
+		{
+			name:   "Invalid platform",
+			source: "",
+			infra:  makeInfrastructureResource(configv1.AWSPlatformType),
+			errMsg: "invalid platform, expected to be OpenStack",
+		}, {
+			name: "Config with unsupported secret-namespace override",
+			source: `[Global]
+secret-namespace = foo
+secret-name = openstack-credentials`,
+			infra:  makeInfrastructureResource(configv1.OpenStackPlatformType),
+			errMsg: "'[Global] secret-namespace' is set to a non-default value",
+		}, {
+			name: "Config with unsupported secret-name override",
+			source: `[Global]
+secret-namespace = kube-system
+secret-name = foo`,
+			infra:  makeInfrastructureResource(configv1.OpenStackPlatformType),
+			errMsg: "'[Global] secret-name' is set to a non-default value",
+		}, {
+			name: "Config with unsupported kubeconfig-path override",
+			source: `[Global]
+secret-namespace = kube-system
+secret-name = openstack-credentials
+kubeconfig-path = https://foo`,
+			infra:  makeInfrastructureResource(configv1.OpenStackPlatformType),
+			errMsg: "'[Global] kubeconfig-path' is set to a non-default value",
+		}, {
+			name:   "Empty config",
+			source: "",
+			infra:  makeInfrastructureResource(configv1.OpenStackPlatformType),
+		}, {
+			name: "Non-empty config",
+			source: `[Global]
+secret-name = openstack-credentials
+secret-namespace = kube-system
+
+[BlockStorage]
```

Comment:
Please can we explicitly test that unknown sections and parameters are passed through unmodified? e.g. add:

```
[blockstorage]
foobarwhizz = 9
```